### DISCUSSION TITLE
Use of legacy rng

### DISCUSSION THREAD

**Discussion by mfry-ll** [yalmip/YALMIP#1082] [Possible bug or missing functionality] [answered]
It appears that yalmip sometimes calls a legacy version of `rng`, which has created some issues when subsequently calling `rng(seed)`.  I can't produce all the code here, but the problem occurs when I do this:
```
rng(0)
% run some tests that use rand
optimize(...)
rng(0) % Error thrown here
```

The error that is thrown:
```
Error using rng (line 99)
The current random number generator is the legacy generator.  This is because you have executed a command such as rand('state',0), which
activates MATLAB's legacy random number behavior.  You may not use RNG to reseed the legacy random number generator.

Use rng('default') to reinitialize the random number generator to its startup configuration, or call RNG using a specific generator type,
such as rng(seed,'twister').
```

I'm not supplying code that define inputs to the `optimize` function here because `optimize` doesn't always cause this error to be thrown, and I haven't found a concise way to force it to. (I can repeatedly make it do so, but the example requires an entirely different [package](https://github.com/iqcToolbox/iqcToolbox), you can see output that shows the failure [here](https://github.com/iqcToolbox/iqcToolbox/runs/5240225810?check_suite_focus=true#step:5:473), I'm 100% certain it is caused by yalmip and not iqcToolbox). I did find a line calling `rand('state', k)` in the yalmip toolbox, but that was not hit before the error is thrown, so I'm not sure where the troublesome call is.

This isn't a blocker for me, because I can change all the calls in iqcToolbox from `rng(seed)` to `rng(seed, 'twister')` and everything is fine, but wanted to raise awareness.

**Comment by johanlofberg** [comment, marked as answer]
Hmm, that is weird. The natural candidate for issues would be gen_rand_hash and rand_hash (which are super core function which I would be extremely hesitant to touch, don't ask me why there are two...). If rng fails, it is coded to switch to a method using rand, with seeding, but for that rand call to happen the rng code should not work elsewhere.

Hence my best bet for a debug would be a break on line 20 in gen_rand_hash and line 28 of rand_hash, to understand why it errors in the rng code and switches to rand.

**Reply by johanlofberg** [reply]
Another long-shot is that it maybe is related to SDPT3? SDPT3 makes calls to rand with seeds in the file HSDsqlpmain

**Reply by mfry-ll** [reply]
Yes, I saw those lines in `gen_rand_hash` and `rand_hash`, but my breaker wasn't hitting them, I don't think they are causing this problem.

Good call on SDPT3, I'll check if the behavior is solver dependent.

**Reply by mfry-ll** [reply]
Indeed, the problem was SDPT3 in `HSDsqlpmain`. My mistake for suggesting yalmip was the problem!

**Reply by mfry-ll** [reply]
A bit puzzling though, since my code ran fine previously, and only just yesterday started to hit this snag...I wonder what changed.

**Reply by johanlofberg** [reply]
I guess you have to figure out when/why HSDsqlpmain is called instead of sqlpmain
